Extend View.cpp unit test coverage.

scottschurr opened the pull request:
A couple of notes on this pull request:

One of the unit tests in particular uses an unusual technique (https://github.com/scottschurr/rippled/blob/9c91b0a676f97380e0e03f69c01eea03c9df4b02/src/ripple/ledger/tests/View_test.cpp#L667-L728).  It holds onto `shared_ptr`s of ledgers past the lifetime of the `Env` that created the ledgers.  This seems to work fine within the confines of this unit test.  I made a cursory examination of one of these orphan ledgers using the debugger and did not spot anything suspicious.  But I can remove that test if folks decide the technique is too risky.

The `GetAmendments` and `DirIsEmpty` tests are slow.  So I put them each in their own test suites.

Additionally, @seelabs consulted with me and suggested a small addition to SConstruct.  That addition is small enough that we felt it could be combined with this pull request.

Reviewers: @nbougalis, @seelabs

vinniefalco commented on src/ripple/ledger/tests/View_test.cpp:
```diff
@@ -524,6 +525,209 @@ class View_test
+
+            env.fund(XRP(10000), alice);
+            env.close();
+            rdViewA3 = env.closed();
```
We should probably audit the code to make sure that ledgers can outlive the `Application` object. Or if not, figure out what the obstacles are. I think `SHAMapFamily` could be one road block.

nbougalis commented:
:+1: from me. As long as Vinnie is satisfied that the issues re: lifetime are OK, this is a go from me.

vinniefalco commented on src/ripple/ledger/tests/View_test.cpp:
```diff
@@ -524,6 +525,209 @@ class View_test
+    {
+        // This test requires incompatible ledgers.  The good news is that
+        // ledgers have their memory managed by shared_ptr.  So we can:
+        //  o Make an Env and construct a couple of ledgers in it,
```
Is `o` a high-ascii character?

scottschurr replied:
Naw, it's a habit of mine from pre-unicode days.  It's a lower case letter 'O'.

vinniefalco replied:
Hmm... `o` `o`

seelabs commented on SConstruct:
```diff
@@ -421,8 +421,8 @@ def get_libs(lib, static):
-    except:
-        raise Exception('pkg-config failed for ' + lib)
+    except Exception as e:
+        raise Exception('pkg-config failed for ' + lib + ' Exception: ' + str(e))
```
Nit: Let's put a semicolon between the messages. So replace `' Exception: '` with `'; Exception: '`.

seelabs commented on src/ripple/ledger/tests/View_test.cpp:
```diff
@@ -524,6 +525,209 @@ class View_test
+            // Alice's USD balance should be zero if frozen.
+            expect(USD(0) == accountHolds (*env.closed(),
+                alice, USD.currency, gw, fhZERO_IF_FROZEN, env.journal));
+
```
There are four interesting scenarios for `accountHolds` and freeze handling:
1) Account Frozen and fhZERO_IF_FROZEN
2) Account Thawed and fhZERO_IF_FROZEN
3) Account Frozen and fhIGNORE_FREEZE
4) Account Thawed and fhIGNORE_FREEZE

We are only testing two here. Let's add the other two.

vinniefalco replied:
The purpose of this test is to exercise functions in `View.cpp` to get coverage up to 100%, not to verify the correct behavior of any particular business logic (except that necessary to get coverage up).

That's not to say that we should not have such tests for freeze, but they belong in a unit test that specifically addresses the verification of correct freeze functionality.

seelabs commented on src/ripple/ledger/tests/View_test.cpp:
```diff
@@ -524,6 +525,209 @@ class View_test
+
+            // Bob's balance should be zero if frozen.
+            expect(USD(0) == accountHolds (*env.closed(),
+                bob, USD.currency, gw, fhZERO_IF_FROZEN, env.journal));
```
See my comment on global freeze. We should be testing the all four combinations for Local freeze and freeze handling.

seelabs commented on src/ripple/ledger/tests/View_test.cpp:
```diff
@@ -524,6 +525,209 @@ class View_test
+        env.fund(XRP(10000), alice, bob, carol, gw);
+        env.trust(USD(100), alice, bob, carol);
+        {
+            // Global freezing.
```
One flag I don't see tested is the `NoFreeze` flag. With this flag the issuing account can no longer enable individual freezes, but can still issue a global freeze. It can not disable a global freeze.

seelabs commented on src/ripple/ledger/tests/View_test.cpp:
```diff
@@ -524,6 +525,209 @@ class View_test
+            // carol's XRP balance should be her holdings minus her reserve.
+            auto const carolsXRP = accountHolds (*env.closed(),
+                carol, xrpCurrency(), gw, fhZERO_IF_FROZEN, env.journal);
+            expect(carolsXRP == XRP(9750));
```
A comment explaining how we got to 9750 here: She has 10000, base reserve is 200, increment is 50 for every item she owns, and she owns one item.

seelabs replied:
I don't like that `gw` is used as the issuer to get the XRP balance. I get that it's ignored, but I'd put the correct issuer there anyway.

seelabs commented on src/ripple/ledger/tests/View_test.cpp:
```diff
@@ -524,6 +525,209 @@ class View_test
+
+            // carol's XRP balance should now show as zero.
+            expect (XRP(0) == accountHolds (*env.closed(),
+                carol, xrpCurrency(), gw, fhZERO_IF_FROZEN, env.journal));
```
Would testing placing an offer for XRP when we are below reserve be out of scope for this PR? I assume yes. If no, then let's also test that we can _only_ place an offer from USD->XRP and not other type of offer.

seelabs commented on src/ripple/ledger/tests/View_test.cpp:
```diff
@@ -524,6 +525,209 @@ class View_test
+            // Gateways have whatever funds they claim to have.
+            auto const gwUSD = accountFunds(
+                *env.closed(), gw, USD(0), fhZERO_IF_FROZEN, env.journal);
+            expect (gwUSD == USD(0));
```
Since the gateway can have whatever they claim to have, we should test a non-zero default as well.

seelabs commented:
:+1: from me. I left some nits you can take or leave as you see fit. Nice tests!

vinniefalco commented:
I'd still like to at least know what the consequences of detached `Ledger` objects are.

scottschurr commented:
@vinniefalco: I agree regarding the detached `Ledger` objects.  Do you have a suggestion for how I should attack the audit?  Even if we decide to throw out the test the results of the audit would be useful to know.

seelabs commented:
@scottschurr I looked for raw pointers and references that would be stale if env went out of scope. I didn't see any, but I could have missed it. I'm doing another pass now.

vinniefalco commented:
@scottschurr I'm not saying to throw out the test results but I would inspect the `Ledger` constructors.

scottschurr commented:
@seelabs.  Nice.  Thanks.  I used the debugger to nose around in one of the detached ledgers.  As I recall I didn't spot any raw pointers.  I saw a number of live `shared_ptr`s and I think I saw one `weak_ptr` that was still live.

scottschurr commented:
@vinniefalco: Right.  I didn't think you were saying we should throw out the test.  You were just expressing appropriate caution about ledgers being used in a new and unexpected circumstance.  I'll dig around a bit more and see if I spot anything risky.

vinniefalco commented:
I would inspect the `Ledger` constructors.

seelabs commented:
Address sanitizer is not running clean on these unittests. It looks like it is unrelated to detached ledgers, but let me look into this.

seelabs commented:
@scottschurr The address sanitizer trigger is unrelated to your changes. @HowardHinnant found the issue. We need to make `Logs` destructor virtual. When I do that address sanitizer detects one memory leak, but it's unrelated to your changes. The good news is the detached ledger runs clean.

scottschurr commented:
@seelabs: Nice.  Thanks for running the address sanitizer.  I need to learn how to do that on OS X.  I'm spending some time right now looking at the data members of a `Ledger` to see if there's anything suspicious.

scottschurr commented:
@vinniefalco: Thanks for pointing me down the right track.  Each `Ledger` holds `shared_ptr`s to two `SHAMap`s.  Each `SHAMap` holds a reference to a `Family` which is held by the `Application`.  When the `Env` destructor runs, those `Family` references refer to released memory.

The obvious places where these stale references could bite us are:
  o `SHAMap::fetchNodeFromDB` -- only a problem if `backed_ == true`
  o `SHAMap::checkFilter` -- could be fixed by using `journal_` instead of `f_.journal()`
  o `SHAMap::descendAsync` -- only a problem if `backed_ == true`
  o `SHAMap::writeNode` -- only a problem if `backed_ == true`
  o ` SHAMap::getCache`
  o `SHAMap::canonicalize`  -- only a problem if `backed_ == true`
  o `SHAMap::getMissingNodes`
  o `SHAMap::addRootNode` -- -- could be fixed by using `journal_` instead of `f_.journal()`
  o `SHAMap::addKnownNode`

`SHAMap` also makes the `Family` public.  These are the places that use the public member:
  o` Ledger::Ledger` -- Around Ledger.cpp line 267.  Create a new open ledger that follows this one
  o `Ledger::visitStateItems`

scottschurr commented:
Addressed comments from @seelabs which didn't involve extending tests.  Modified the compatible ledger test to keep two live `Env` instances in order to build incompatible ledgers.

seelabs commented:
:+1: I found a minor formatting nit you should probably fit, but other than that looks good.

scottschurr commented:
Fixed formatting.  Thanks @seelabs.

nbougalis commented:
:+1: from me

scottschurr commented:
Rebased to -b9.

vinniefalco commented:
:+1:
